design/one-pager-stacks-security-isolation.md

```diff
@@ -241,13 +235,134 @@ permissionScope: Cluster
+ infrastructure persona, whereas any namespace stack resources will be for the namespace that that stack is installed
+ into, thus the application persona.
+
+At install time:
+ - Cluster Stack: generate cluster roles for infra persona that auto-aggregate to built-in roles for
+ infrastructure owners.
+ - Namespaced Stack: generate cluster roles for the workspace persona different permission
 
-With these two separate types, an administrator can allow access to installing cluster stacks to only certain restricted roles, then perhaps be more permissive with access to installing namespaced stacks.
+#### Built in ClusterRoles
+
+The following built in roles will be installed and packaged with Crossplane, and updated with role aggregation as we
+ install additional stacks.
+
+The verbs supported in these roles(admin,edit,view) will model based on existing logic around
+[built-in kubernetes roles](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#default-roles-and-role-bindings).
+
+|ClusterRole Name | Permissions |
+|:-------|:----------|
+| crossplane:admin | Admin: environment, workspaces(namespaces), ClusterRole, Role, ClusterRoleBinding, RoleBinding |
+| crossplane:infra-admin | Admin: environment, manage cluster stacks |
+| crossplane:infra-edit  | Edit: environment, manage cluster stacks |
+| crossplane:infra-viewer | View: environment. Not permitted: secrets |
+
+If we create a workspace(specially annotated+labeled namespace), the stack-manager will create the corresponding cluster
+ roles for that workspace. This is for the benefit of external tools, that could role bind to this known clusterRole
+ that will be kept up to date automatically with ClusterRole Aggregation as the stack-manager installs and uninstalls
+ new stacks.
+
+ crossplane:ns-name:workspace-admin
+ crossplane:ns-name:workspace-edit
+ crossplane:ns-name:workspace-view
+
+To grant jane user admin access to the foo workspace, the following yaml would suffice
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+# This role binding allows "jane" to admin the Crossplane foo workspace.
+kind: RoleBinding
+metadata:
+  name: admin-workspace
+  namespace: foo
+subjects:
+- kind: User
+  name: jane # Name is case sensitive
+  apiGroup: rbac.authorization.k8s.io
+roleRef:
+  kind: ClusterRole
+  name: crossplane:foo:workspace-admin
+  apiGroup: rbac.authorization.k8s.io
+```
 
-If the install scope type does not match what the stack author has declared within their stack metadata, an error status should be set on the install resource and installation of the stack should not proceed.
+#### Managing CRD Permissions of Dynamically (Aggregation)
+
+As we install new stacks and new CRDs we will need to grant our built in roles access to delete, edit, create, and view
+ the given newly installed resources, and we can use the built in
+ [aggregated cluster roles](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#aggregated-clusterroles)
```

Comment:
@lukeweber it looks like aggregated roles only work for cluster roles, so for workspace roles we'd need to create additional role binding for each stacks, or it looks like you're proposing `ClusterRole` definitions would be created for Namespace Stacks as well?

Reply:
ClusterRoles for every stack was this proposal. Still considering the implications of which you’ve surfaced a few in this doc. Maybe it’s not the right approach, but was admittedly stretching the previous aggregation idea to its limits to see what it looked like.

Reply:
https://kubernetes.io/docs/reference/access-authn-authz/rbac/

> A RoleBinding may also reference a ClusterRole to grant the permissions to namespaced resources defined in the ClusterRole within the RoleBinding’s namespace. This allows administrators to define a set of common roles for the entire cluster, then reuse them within multiple namespaces.

In support of the notion that a stack's CRDs could be installed once (upon first StackInstall) and then multiple instances of the "Stack" (or StackInstall) could make that stack's controller available within new namespaces, this ClusterRole could be reusable.

The ClusterRole could be reused when creating the `:system` rolebinding needed for the controller.  The `system` roles could change between Stack versions, so we wouldn't want to reuse clusterroles between versions - this aligns with this design.

When creating RoleBindings for workspace admins, we should still be able to benefit from ClusterRole aggregation, allowing a workspace admin to administer any (and only) the Stacks installed in the namespace where their (namespaced) RoleBinding (of a workspace clusterrole) was bound.

This sounds right to me.